tb_plugin/torch_tb_profiler/profiler/memory_parser.py

```diff
@@ -0,0 +1,212 @@
+import os
+from collections import defaultdict
+
+from .. import utils
+from .node import MemoryMetrics, is_operator_node
+from .trace import DeviceType, EventTypes
+
+logger = utils.get_logger()
+
+BENCHMARK_MEMORY = os.getenv('TORCH_PROFILER_BENCHMARK_MEMORY')
+if BENCHMARK_MEMORY is not None and BENCHMARK_MEMORY.upper() in ("1", "TRUE", "ON"):
+    BENCHMARK_MEMORY = True
+else:
+    BENCHMARK_MEMORY = False
+
+class MemoryRecord:
+    def __init__(self, scope, pid, tid, ts, device_type, device_id, bytes):
+        self.scope = scope
+        self.tid = tid
+        self.pid = pid
+        self.ts = ts
+        self.device_type = device_type
+        self.device_id = device_id
+        self.bytes = bytes
+
+    @property
+    def device_name(self):
+        if self.device_type == DeviceType.CPU:
+            return "CPU"
+        elif self.device_type == DeviceType.CUDA:
+            return "GPU{}".format(self.device_id)
+        else:
+            return None
+
+
+class MemoryParser:
+    def __init__(self, tid2tree):
+        self.tid2tree = tid2tree
+
+        self.records_by_tid = defaultdict(list)
+
+        # statistics purpose
+        self.staled_records = []
+        self.processed_records = []
+
+    def parse_events(self, events):
+        for event in events:
+            if event.type == EventTypes.MEMORY:
+                record = MemoryRecord(event.scope, event.pid, event.tid, event.ts, event.device_type, event.device_id, event.bytes)
+                self.records_by_tid[record.tid].append(record)
+
+        for val in self.records_by_tid.values():
+            val.sort(key=lambda x: x.ts)
+
+        if BENCHMARK_MEMORY:
+            import time
+            start = time.time_ns()
+            self.update_node()
+            end = time.time_ns()
+            print("{}: update node takes: {}".format(os.getpid(), end - start))
+
+            start = time.time_ns()
+            self.update_node_loop()
+            end = time.time_ns()
+            print("{}: update node using loops takes: {}".format(os.getpid(), end - start))
```

Comment:
Also compare the 2 solutions' result data, to validate correctness?

Reply:
yes, they are same.